### PR TITLE
Replace usages of bytes.Buffer with strings.Builder

### PR DESCRIPTION
**Pull request by mattrobenolt**
## Description

I went through each of these manually and eyeballed to make sure they followed the pattern of only creating the buffer and using the result of .String() after later.

I didn't do everything within tests since I didn't think it was too relevant.

I also unified some of the usage patterns to be `var buf strings.Builder` as opposed to `buf := new(bytes.Buffer)` etc.

All of the rest of the changes were just from gofmt. I'm assuming they are fine, but I understand this adds some more noise to the diffs and might not be desired. I can look into teasing them out if that's a concern.

But overall, this change should yield 1 less memory allocation since the bytes -> string is done without an extra allocation.

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/bytes2/buffer.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ func (buf *Buffer) String() string {
 // is _not_ allocated, so modifying this buffer after calling StringUnsafe will lead
 // to undefined behavior.
 func (buf *Buffer) StringUnsafe() string {
-	return *(*string)(unsafe.Pointer(&buf.bytes))
+	return unsafe.String(unsafe.SliceData(buf.bytes), len(buf.bytes))
```
Comment (mattrobenolt) [inline go/bytes2/buffer.go:68]: I can pull this out into a separate change, but just came across it while doing all of this, and figured it'd be good to adopt the newer convention from go1.20.

File: go/sqltypes/value_test.go [hunk 2]
```diff
@@ -463,13 +463,13 @@ func TestEncode(t *testing.T) {
...
-		buf := &bytes.Buffer{}
-		tcase.in.EncodeSQL(buf)
+		var buf strings.Builder
+		tcase.in.EncodeSQL(&buf)
 		if tcase.outSQL != buf.String() {
 			t.Errorf("%v.EncodeSQL = %q, want %q", tcase.in, buf.String(), tcase.outSQL)
 		}
-		buf = &bytes.Buffer{}
-		tcase.in.EncodeASCII(buf)
+		buf.Reset()
```
Comment (mattrobenolt) [inline go/sqltypes/value_test.go:471]: Changed to just reset and reuse the buffer instead of creating a new one.

File: go/vt/servenv/status.go [hunk 3]
```diff
@@ -301,7 +300,7 @@ func registerDebugBlockProfileRate() {
 		runtime.SetBlockProfileRate(rate)
 		log.Infof("Set block profile rate to: %d", rate)
 		w.Header().Set("Content-Type", "text/plain")
-		w.Write([]byte(message))
+		io.WriteString(w, message)
```
Comment (mattrobenolt) [inline go/vt/servenv/status.go:303]: Other places within here were using `io.WriteString`, so maintained the convention.

File: go/vt/vtexplain/vtexplain.go [hunk 4]
```diff
@@ -154,10 +151,11 @@ type (
 func (tq *TabletQuery) MarshalJSON() ([]byte, error) {
 	// Convert Bindvars to strings for nicer output
 	bindVars := make(map[string]string)
+	var buf strings.Builder
 	for k, v := range tq.BindVars {
-		var b strings.Builder
-		sqlparser.EncodeValue(&b, v)
-		bindVars[k] = b.String()
+		buf.Reset()
```
Comment (mattrobenolt) [inline go/vt/vtexplain/vtexplain.go:156]: Might as well reuse the buffer here.

**Review by vmg — APPROVED** [review]
These are all reasonable changes. Thanks Matt!

**Review by deepthi — APPROVED** [review]
Nice work! I was wondering if we could run micro benchmarks that will show an improvement, but since these changes are all over the code base, it's probably not worth trying to run any of the ones we _do_ have - like the sqlparser ones.

**Comment by mattrobenolt** [comment]
> Nice work! I was wondering if we could run micro benchmarks that will show an improvement, but since these changes are all over the code base, it's probably not worth trying to run any of the ones we do have - like the sqlparser ones.

Yeah, I considered that, but figured it wasn't really worth it. It's likely an impact, but hard to quantify if it's measurable or not and would depend on the workload, etc.

**Pull request merged**